File: week5&6/3_is_bst_advanced/is_bst_hard.cpp
#include <algorithm>
#include <iostream>
#include <map>


using std::cin;
using std::cout;
using std::endl;
using std::map;
using std::ios_base;

struct Node {
    int key;
    int left;
    int right;

    Node() : key(0), left(-1), right(-1) {}

    Node(int key_, int left_, int right_) : key(key_), left(left_), right(right_) {}
};

bool IsBinarySearchTree( map<int, Node> &tree, const int node, const int min, const int max) {
    if (tree.find(node) == tree.end())
        return true;
    if ((tree[node].key < min) || (tree[node].key > max))
        return false;
    return IsBinarySearchTree(tree, tree[node].left, min, tree[node].key - 1) &&
    IsBinarySearchTree(tree, tree[node].right, tree[node].key, max);
}

int main() {
//    ios_base::sync_with_stdio(false);
    int nodes;
    cin >> nodes;
    map<int, Node> tree;
    int min = -1;
    int max = std::numeric_limits<int>::max();
    for (int i = 0; i < nodes; ++i) {
        int key, left, right;
        cin >> key >> left >> right;
        tree[i] = Node(key, left, right);
    }
    if (IsBinarySearchTree(tree, 0, min, max)) {
        cout << "CORRECT" << endl;
    } else {
        cout << "INCORRECT" << endl;
    }
    return 0;
}

File: week5&6/3_is_bst_advanced/is_bst_hard.py
#!/usr/bin/python3

import sys
import threading
import math

sys.setrecursionlimit(10 ** 7)  # max depth of recursion
# had to increase stack to ** 27 from ** 25
threading.stack_size(2 ** 27)  # new thread will get stack of such size

tree = {}


def is_binary_search_tree(node, neg, pos):
    if node not in tree:
        return True
    # if current key is smaller than neg(left child) or larger than pos(right child)
    if tree[node][0] < neg or tree[node][0] > pos:
        return False
    # run on left and right children
    return is_binary_search_tree(tree[node][1], neg, tree[node][0] - 1) and \
           is_binary_search_tree(tree[node][2], tree[node][0], pos)


def main():
    nodes = int(sys.stdin.readline().strip())
    # the tree dictionary will include list of 3 elements each
    global tree
    neg_inf = -math.inf
    pos_inf = math.inf
    for i in range(nodes):
        tree[i] = list(map(int, sys.stdin.readline().strip().split()))
    if is_binary_search_tree(0, neg_inf, pos_inf):
        print("CORRECT")
    else:
        print("INCORRECT")


threading.Thread(target=main).start()

'''
7
4 1 2
2 3 4
6 5 6
1 -1 -1
3 -1 -1
5 -1 -1
7 -1 -1
'''